src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class CSharpDeterministicKeyBuilder : DeterministicKeyBuilder
+    {
+        internal static readonly CSharpDeterministicKeyBuilder Instance = new();
+
+        private CSharpDeterministicKeyBuilder()
+        {
+
+        }
+
+        protected override void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            if (options is not CSharpCompilationOptions csharpOptions)
+            {
+                throw new InvalidOperationException();
+            }
+
+            base.WriteCompilationOptionsCore(writer, options);
+
+            writer.Write("unsafe", csharpOptions.AllowUnsafe);
+            writer.Write("topLevelBinderFlags", csharpOptions.TopLevelBinderFlags);
+            writer.WriteKey("globalUsings");
+            writer.WriteArrayStart();
+            foreach (var name in csharpOptions.Usings)
+            {
+                writer.Write(name);
+            }
+            writer.WriteArrayEnd();
+        }
+
+        protected override void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
+        {
+            if (parseOptions is not CSharpParseOptions csharpOptions)
```

Comment:
make DeterministicKeyBuilder  generic on TParseOptions and this can be moved to base type and derived types can always get the type they expect.

Reply:
I experimented with this and the end result is that either: 

1. The `public` API point becomes generic because the generic ends up propogating there through `CreateDeterministicBuilder` 
2. Introduce another level in the hierarchy which is non-generic that we can continue to expose

Reply:
I chased this idea for a little while and I don't think the result makes the code cleaner. The pattern I ended up with is essentially the following: 

```c#
internal abstract class DeterministicKeyBuilder
{

    protected abstract WriteParseOptions(JsonWriter writer, ParseOptions parseOptions);
}


internal abstract class DeterministicKeyBuilder<TParseOptions, TCompilationOptions> : DeterministicKeyBuilder
    where TParseOptions : ParseOptions
    where TCompilationOptions : CompilationOptions
{

    protected override void WriteParseOptions(JsonWriter writer, ParseOptions parseOptions)
    {
        writer.WriteObjectStart();
        if (parseOptions is not TParseOptions derivedParseOptions)
        {
            throw new ArgumentException(null, nameof(parseOptions));
        }

        WriteParseOptionsCore(writer, derivedParseOptions);
        writer.WriteObjectEnd();
    }
}
```

The non-generic `DeterministicKeyBuilder` is necessary because `CompilationOptions` can't expose a generic type here. Hence you end up with the exact same pattern we were trying to avoid by adding another level in the hierarchy. Didn't feel like a great trade off. 


Let me know if you think I missed an easier path here.

Reply:
nope!  i'm fine with you doing this.